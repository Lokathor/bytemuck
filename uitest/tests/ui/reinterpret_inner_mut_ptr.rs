use bytemuck::ReinterpretInner;

fn main() {
  let _: *mut u16 = (&mut [0u8; 2] as *mut [u8; 2]).reinterpret_inner();
  let _: *mut u32 = (&mut [0u16; 2] as *mut [u16; 2]).reinterpret_inner();
  let _: *mut u64 = (&mut [0u32; 2] as *mut [u32; 2]).reinterpret_inner();

  let _: *mut [u8; 2] = (&mut 0u8 as *mut u8).reinterpret_inner();
  let _: *mut [u16; 2] = (&mut 0u16 as *mut u16).reinterpret_inner();
  let _: *mut [u64; 2] = (&mut 0u64 as *mut u64).reinterpret_inner();
}
