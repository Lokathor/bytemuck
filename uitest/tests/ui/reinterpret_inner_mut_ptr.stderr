error[E0080]: evaluation of `<*const [u8; 2] as bytemuck::cast::CastRaw<*const u16>>::ASSERT` failed
 --> $WORKSPACE/src/cast.rs
  |
  | /     assert!(
  | |       align_of::<Src>() >= align_of::<Dst>(),
  | |       "Attempted conversion to a type with a stricter alignment"
  | |     );
  | |_____^ the evaluated program panicked at 'Attempted conversion to a type with a stricter alignment', $WORKSPACE/src/cast.rs:824:5
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: erroneous constant used
   --> $WORKSPACE/src/cast.rs
    |
    |   const ASSERT: () = <*const Src as CastRaw<*const Dst>>::ASSERT;
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

note: the above error was encountered while instantiating `fn <*mut [u8; 2] as bytemuck::ReinterpretInner<'_, *mut u16>>::reinterpret_inner`
 --> tests/ui/reinterpret_inner_mut_ptr.rs:4:21
  |
4 |   let _: *mut u16 = (&mut [0u8; 2] as *mut [u8; 2]).reinterpret_inner();
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0080]: evaluation of `<*const [u16; 2] as bytemuck::cast::CastRaw<*const u32>>::ASSERT` failed
 --> $WORKSPACE/src/cast.rs
  |
  | /     assert!(
  | |       align_of::<Src>() >= align_of::<Dst>(),
  | |       "Attempted conversion to a type with a stricter alignment"
  | |     );
  | |_____^ the evaluated program panicked at 'Attempted conversion to a type with a stricter alignment', $WORKSPACE/src/cast.rs:824:5
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn <*mut [u16; 2] as bytemuck::ReinterpretInner<'_, *mut u32>>::reinterpret_inner`
 --> tests/ui/reinterpret_inner_mut_ptr.rs:5:21
  |
5 |   let _: *mut u32 = (&mut [0u16; 2] as *mut [u16; 2]).reinterpret_inner();
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0080]: evaluation of `<*const [u32; 2] as bytemuck::cast::CastRaw<*const u64>>::ASSERT` failed
 --> $WORKSPACE/src/cast.rs
  |
  | /     assert!(
  | |       align_of::<Src>() >= align_of::<Dst>(),
  | |       "Attempted conversion to a type with a stricter alignment"
  | |     );
  | |_____^ the evaluated program panicked at 'Attempted conversion to a type with a stricter alignment', $WORKSPACE/src/cast.rs:824:5
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn <*mut [u32; 2] as bytemuck::ReinterpretInner<'_, *mut u64>>::reinterpret_inner`
 --> tests/ui/reinterpret_inner_mut_ptr.rs:6:21
  |
6 |   let _: *mut u64 = (&mut [0u32; 2] as *mut [u32; 2]).reinterpret_inner();
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0080]: evaluation of `<*const u8 as bytemuck::cast::CastRaw<*const [u8; 2]>>::ASSERT` failed
 --> $WORKSPACE/src/cast.rs
  |
  | /     assert!(
  | |       size_of::<Src>() == size_of::<Dst>(),
  | |       "Attempted conversion between types of different sizes."
  | |     );
  | |_____^ the evaluated program panicked at 'Attempted conversion between types of different sizes.', $WORKSPACE/src/cast.rs:820:5
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn <*mut u8 as bytemuck::ReinterpretInner<'_, *mut [u8; 2]>>::reinterpret_inner`
 --> tests/ui/reinterpret_inner_mut_ptr.rs:8:25
  |
8 |   let _: *mut [u8; 2] = (&mut 0u8 as *mut u8).reinterpret_inner();
  |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0080]: evaluation of `<*const u16 as bytemuck::cast::CastRaw<*const [u16; 2]>>::ASSERT` failed
 --> $WORKSPACE/src/cast.rs
  |
  | /     assert!(
  | |       size_of::<Src>() == size_of::<Dst>(),
  | |       "Attempted conversion between types of different sizes."
  | |     );
  | |_____^ the evaluated program panicked at 'Attempted conversion between types of different sizes.', $WORKSPACE/src/cast.rs:820:5
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn <*mut u16 as bytemuck::ReinterpretInner<'_, *mut [u16; 2]>>::reinterpret_inner`
 --> tests/ui/reinterpret_inner_mut_ptr.rs:9:26
  |
9 |   let _: *mut [u16; 2] = (&mut 0u16 as *mut u16).reinterpret_inner();
  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0080]: evaluation of `<*const u64 as bytemuck::cast::CastRaw<*const [u64; 2]>>::ASSERT` failed
 --> $WORKSPACE/src/cast.rs
  |
  | /     assert!(
  | |       size_of::<Src>() == size_of::<Dst>(),
  | |       "Attempted conversion between types of different sizes."
  | |     );
  | |_____^ the evaluated program panicked at 'Attempted conversion between types of different sizes.', $WORKSPACE/src/cast.rs:820:5
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn <*mut u64 as bytemuck::ReinterpretInner<'_, *mut [u64; 2]>>::reinterpret_inner`
  --> tests/ui/reinterpret_inner_mut_ptr.rs:10:26
   |
10 |   let _: *mut [u64; 2] = (&mut 0u64 as *mut u64).reinterpret_inner();
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
